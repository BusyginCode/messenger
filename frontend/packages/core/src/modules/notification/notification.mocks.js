export const notification = {
  notifications: [],
  isLoaded: false
}
