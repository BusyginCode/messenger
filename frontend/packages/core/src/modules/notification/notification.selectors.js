export const getNotifications = state => state.notification.notifications
