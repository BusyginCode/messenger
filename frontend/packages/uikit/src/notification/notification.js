import React from 'react'
import PropTypes from 'prop-types'
import cn from 'classnames'
import Box from '../box'
import Text from '../text'
import Button from '../button'
import styles from './notification.scss'

const Notification = props => {
  const { title, subTitle, icon, subIcon, type, isRead, button, ...rest } = props
  const hasSub = subTitle || subIcon

  return (
    <Box padding="m" className={cn(styles.general, styles[type], { [styles.isRead]: isRead })}>
      {icon && (
        <Box className={styles.icon} right="m">
          {icon}
        </Box>
      )}
      <Box justify="between" align="start" flexGrow={1} {...rest}>
        <div>
          <Text size="m">{title}</Text>
          {hasSub && (
            <Box top="s">
              {subIcon && (
                <Box className={styles.subIcon} right="s">
                  {subIcon}
                </Box>
              )}
              {subTitle && <Text size="s">{subTitle}</Text>}
            </Box>
          )}
        </div>
        {button && (
          <Box left="s" flexShrink={0}>
            <Button size="m" {...button} />
          </Box>
        )}
      </Box>
    </Box>
  )
}

Notification.propTypes = {
  icon: PropTypes.node,
  iconType: PropTypes.string,
  type: PropTypes.string,
  isRead: PropTypes.bool,
  subIcon: PropTypes.node,
  title: PropTypes.node,
  subTitle: PropTypes.node,
  button: PropTypes.object
}

Notification.defaultProps = {
  type: 'default'
}

export default Notification
