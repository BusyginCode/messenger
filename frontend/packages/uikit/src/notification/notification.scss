.general {
  cursor: pointer;
  border-bottom: var(--b-border-width-m) solid var(--b-color-body4);

  &.default {
    background-color: var(--b-color-nav3);

    &.isRead {
      background-color: var(--b-color-body6);
    }
  }

  &.system {
    background-color: var(--b-color-body);
    color: var(--b-color-nav0);
    cursor: default;
    border-bottom: var(--b-border-width-m) solid var(--b-color-body1);
  }
}

.subIcon {
  color: var(--b-color-body3);
}
