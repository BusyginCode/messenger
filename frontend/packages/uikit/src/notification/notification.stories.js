import React from 'react'
import { storiesOf } from '@storybook/react'
import Notification from './notification'

storiesOf('RFC/UIKit/Notification', module).add('default', () => (
  <div style={{ width: 500 }}>
    <Notification title="Hello" canClose onClose={() => {}} />
  </div>
))
