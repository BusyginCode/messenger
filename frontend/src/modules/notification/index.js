export * from './notification.actions'
export * from './notification.selectors'

export { default } from './notification.reducer'
