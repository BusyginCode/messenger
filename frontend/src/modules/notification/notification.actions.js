import { createAction } from 'redux-actions'
import uniqueId from 'lodash/uniqueId'

export const addNotification = createAction('NOTIFICATIONS_ADD')
export const hide = createAction('NOTIFICATIONS_HIDE')

export const show = notification => {
  return ({ dispatch }) => {
    let { duration, id = uniqueId('notification_'), ...data } = notification
    if (data.type === 'toast') {
      duration = duration || 5000
    }

    const uniqueNotification = {
      ...data,
      id
    }

    dispatch(addNotification(uniqueNotification))

    if (typeof duration === 'number') {
      setTimeout(() => {
        dispatch(hide(id))
      }, duration)
    }
  }
}
