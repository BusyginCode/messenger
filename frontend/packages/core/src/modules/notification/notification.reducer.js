import { handleActions } from 'redux-actions'
import { addNotification, hide } from './notification.actions'

export const initialState = {
  notifications: [],
  isLoaded: false
}

const handlerAddNotification = (state, { payload }) => ({
  ...state,
  notifications: [...state.notifications.filter(n => n.id !== payload.id), payload]
})

const handlerClose = (state, { payload }) => ({
  ...state,
  notifications: state.notifications.filter(
    payload instanceof Array ? n => !payload.includes(n.id) : n => n.id !== payload
  )
})

const reducer = handleActions(
  {
    [addNotification]: handlerAddNotification,
    [hide]: handlerClose
  },
  initialState
)

export default reducer
