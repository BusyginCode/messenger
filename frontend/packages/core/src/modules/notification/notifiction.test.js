import { createStore } from '@messenger/testing'
import { addNotification, show, hide } from './notification.actions'
import { getNotifications } from './notification.selectors'
import notificationReducer from './notification.reducer'

jest.mock('lodash/uniqueId', () => {
  return () => 'ID'
})

const uniqueNotification = {
  id: 'ID',
  message: 'message',
  type: 'alert',
  page: true
}

describe('notifications actions', () => {
  const { dispatch, getState } = createStore({
    notification: notificationReducer
  })

  it('should add notification correctly', () => {
    expect(dispatch(addNotification(uniqueNotification))).toMatchSnapshot()
    expect(getState()).toMatchSnapshot()
  })

  it('should hide correctly', () => {
    expect(dispatch(hide('ID'))).toMatchSnapshot()
    expect(getState()).toMatchSnapshot()
  })

  describe('show', () => {
    it('adds notification without timeout when duration is not provided', () => {
      const notification = {
        message: 'message',
        type: 'alert',
        page: true
      }

      const dispatch = jest.fn()
      global.setTimeout = jest.fn()

      show(notification)({ dispatch })

      expect(dispatch).toHaveBeenCalledWith({
        payload: uniqueNotification,
        type: 'NOTIFICATIONS_ADD'
      })

      expect(global.setTimeout).not.toHaveBeenCalled()
    })

    it('hide notification with "toast" type with default timeout when duration is not provided', () => {
      const notification = {
        message: 'message',
        type: 'toast',
        page: true
      }

      const dispatch = jest.fn()
      global.setTimeout = jest.fn()

      show(notification)({ dispatch })

      expect(global.setTimeout).toHaveBeenCalledWith(expect.any(Function), 5000)
    })

    it('adds notification with timeout when duration is provided', () => {
      const duration = 1000
      const notification = {
        message: 'message',
        type: 'alert',
        page: true,
        duration
      }

      const dispatch = jest.fn()
      global.setTimeout = jest.fn(func => func())

      show(notification)({ dispatch })

      expect(dispatch).toHaveBeenCalledWith({
        payload: uniqueNotification,
        type: 'NOTIFICATIONS_ADD'
      })

      expect(global.setTimeout).toHaveBeenCalledWith(expect.any(Function), duration)

      expect(dispatch).toHaveBeenCalledWith({
        payload: 'ID',
        type: 'NOTIFICATIONS_HIDE'
      })
    })
  })
})

describe('notification selectors', () => {
  it('should getNotifications correctly', () => {
    const notifications = [1, 2, 3]
    expect(getNotifications({ notification: { notifications } })).toBe(notifications)
  })
})

describe('notification reducer', () => {
  it('should reduce initialState correctly', () => {
    expect(notificationReducer(undefined, {})).toMatchSnapshot()
  })

  it('should reduce [addNotification] correctly', () => {
    expect(notificationReducer(undefined, addNotification(uniqueNotification))).toMatchSnapshot()
  })

  it('should reduce [hide] correctly', () => {
    expect(notificationReducer({ notifications: [{ id: 'a' }] }, hide('a'))).toMatchSnapshot()
  })

  it('should reduce [hide] array correctly', () => {
    expect(
      notificationReducer(
        {
          notifications: [{ id: 'a' }, { id: 'b' }]
        },
        hide(['a', 'b'])
      )
    ).toMatchSnapshot()
  })
})
