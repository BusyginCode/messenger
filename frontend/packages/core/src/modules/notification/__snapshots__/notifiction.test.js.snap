// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notification reducer should reduce [addNotification] correctly 1`] = `
Object {
  "isLoaded": false,
  "notifications": Array [
    Object {
      "id": "ID",
      "message": "message",
      "page": true,
      "type": "alert",
    },
  ],
}
`;

exports[`notification reducer should reduce [hide] array correctly 1`] = `
Object {
  "notifications": Array [],
}
`;

exports[`notification reducer should reduce [hide] correctly 1`] = `
Object {
  "notifications": Array [],
}
`;

exports[`notification reducer should reduce initialState correctly 1`] = `
Object {
  "isLoaded": false,
  "notifications": Array [],
}
`;

exports[`notifications actions should add notification correctly 1`] = `
Object {
  "payload": Object {
    "id": "ID",
    "message": "message",
    "page": true,
    "type": "alert",
  },
  "type": "NOTIFICATIONS_ADD",
}
`;

exports[`notifications actions should add notification correctly 2`] = `
Object {
  "notification": Object {
    "isLoaded": false,
    "notifications": Array [
      Object {
        "id": "ID",
        "message": "message",
        "page": true,
        "type": "alert",
      },
    ],
  },
}
`;

exports[`notifications actions should hide correctly 1`] = `
Object {
  "payload": "ID",
  "type": "NOTIFICATIONS_HIDE",
}
`;

exports[`notifications actions should hide correctly 2`] = `
Object {
  "notification": Object {
    "isLoaded": false,
    "notifications": Array [],
  },
}
`;
